### PR TITLE
[ll] Fix garbage leak and implement allocation for render

### PR DESCRIPTION
**Pull request by Bastacyclop**
Working on #1484.
With the simplest `BoxedAllocator` (allocates an heap for each resource) the example looks fine.
With the `StackAllocator` the logo is not displayed, do I need to use the heap offsets when doing mapping, copies, .. ?
@kvark I hope that's simple enough ;)

### CODE REVIEW

File: examples/render/quad_render/main.rs [hunk 1]
```diff
@@ -216,34 +216,25 @@ fn main() {
...
-    let vertex_buffer = {
-        let buffer = device.create_buffer(buffer_len, buffer_stride, buffer::VERTEX).unwrap();
-        println!("{:?}", buffer);
-        device.bind_buffer_memory(&heap, 0, buffer).unwrap()
-    };
-
-    // TODO: check transitions: read/write mapping and vertex buffer read
-    device.write_mapping::<Vertex>(&vertex_buffer, 0..buffer_len)
+    let vertex_buffer = context.mut_device().create_buffer::<Vertex, _>(
+        &mut allocator,
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply examples/render/quad_render/main.rs]: Will do.

File: examples/render/quad_render/main.rs [hunk 2]
```diff
@@ -216,34 +216,25 @@ fn main() {
...
+        gfx::buffer::Role::Vertex,
+        gfx::memory::Usage::Upload,
+        gfx::memory::Bind::empty(),
+        QUAD.len() as u64
+    ).unwrap();
+
+    device.write_mapping::<Vertex>(vertex_buffer.resource(), 0..buffer_len)
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply examples/render/quad_render/main.rs]: Yes I use this temporary example as some sort of integration process.

File: src/render/src/memory.rs [hunk 3]
```diff
@@ -131,3 +187,40 @@ impl<I, T> ops::DerefMut for Typed<I, T> {
         &mut self.inner
     }
 }
+
+/// This is the unique owner of the inner struct.
+#[derive(Debug)]
+pub struct DropDelayed<T>(Arc<UnsafeCell<T>>);
+/// Keep-alive without any access (only Drop if last one).
+pub struct DropDelayer<T>(Arc<UnsafeCell<T>>);
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/render/src/memory.rs]: Could be `Provider` and `Dependency`.

**Comment by msiglreith** [comment]
Looks good so far, beside the points mentioned by kvark.
Would you mind explaining the reason for the garbage leak and your fix?

**Comment by Bastacyclop** [comment]
@msiglreith There is nothing to prevent the handles from outliving the `Context`, such handles would `Drop` after the [final cleanup](https://github.com/gfx-rs/gfx/pull/1492/files#diff-be2c920c08146924af4d1983d6894d63L476) and [send on a disconnected channel](https://github.com/Bastacyclop/gfx/blob/d59de7a9c5dd30efe952548ae2c1d65a54b2f667/src/render/src/handle.rs#L147). I have not observed this in the example but it could happend pretty easily. The fix is to [keep the garbage collector alive](https://github.com/Bastacyclop/gfx/blob/d59de7a9c5dd30efe952548ae2c1d65a54b2f667/src/render/src/handle.rs#L21) until all handles are dropped using the [`Provider`/`Dependency`](https://github.com/Bastacyclop/gfx/blob/d59de7a9c5dd30efe952548ae2c1d65a54b2f667/src/render/src/memory.rs#L185) abstraction extracted from the `Pool`/ `PoolDependency` implementation.

File: src/core/src/buffer.rs [hunk 4]
```diff
@@ -64,3 +64,23 @@ pub struct IndexBufferView<'a, B: Backend> {
...
+    buffer: &B::UnboundBuffer,
+    usage: Usage,
+) -> memory::Requirements {
+    use std::cmp::max;
+    use device::Device;
+
+    let mut requirements = device.get_buffer_requirements(buffer);
+    if usage.can_transfer() {
+        let limits = device.get_limits();
+        requirements.alignment = max(
```
Comment (kvark) [inline src/core/src/buffer.rs]: holy greatness :)

**Pull request merged**